### PR TITLE
Add autogenerated documentation (with mild refactoring)

### PR DESCRIPTION
**Pull request by TariqAHassan**
This PR mainly aims to add autogenerated documentation via Sphinx.
In order for Sphinx to be able to understand the docstrings that were present,
they had to be slighly restructured (I think a hybrid of the Google style
and reStructured text was being used, so I just chose to standardize against the former).

Notes: 
  * I have included some documentation on how to generate the documentation in `docs/README.md`.
  * the documentation is rendered to a (new) gh-pages branch (see #81).

Additionally, I:
   * updated the `setup.py` file to ensure that all dependencies are installed.
   * renamed `EuclidianSimilarityPredictionGraph` to `EuclideanSimilarityPredictionGraph` (fixing a typo). However, I added an alias to the old name (`EuclidianSimilarityPredictionGraph`) to ensure no code breaks.

Thanks for making this cool package : )

### CODE REVIEW

**Comment by jfkirk** [comment]
Hey @TariqAHassan -- this is tremendous, thank you! This is a big change, so it will take some time for me to fully review and wrap my head around.

I'm very excited about this change, and deeply appreciative of your contribution!

File: tensorrec/prediction_graphs.py [hunk 1]
```diff
@@ -9,80 +15,123 @@ class AbstractPredictionGraph(object):
...
         gathered_user_reprs = tf.gather(normalized_users, tf_x_user)
         gathered_item_reprs = tf.gather(normalized_items, tf_x_item)
         return tf.reduce_sum(tf.multiply(gathered_user_reprs, gathered_item_reprs), axis=1)
 
 
-class EuclidianSimilarityPredictionGraph(AbstractPredictionGraph):
-    """
-    This prediction function calculates the prediction as the negative euclidian distance between the user and
-    item representations.
+class EuclideanSimilarityPredictionGraph(AbstractPredictionGraph):
```
Comment (jfkirk) [inline tensorrec/prediction_graphs.py]: 🤦‍♂️  Thanks!

File: tensorrec/prediction_graphs.py [hunk 2]
```diff
@@ -115,3 +174,8 @@ def connect_serial_prediction_graph(self, tf_user_representation, tf_item_repres
         distance = tf.maximum(distance, self.epsilon)
 
         return -1.0 * tf.sqrt(distance)
+
+
+# ToDo: depreciate
+# Alias to the former class name to ensure backwards compatibility.
+EuclidianSimilarityPredictionGraph = EuclideanSimilarityPredictionGraph
```
Comment (jfkirk) [inline tensorrec/prediction_graphs.py]: Perfect. On the next major release, we can strip this out.

File: requirements.txt [hunk 3]
```diff
@@ -1,3 +1,6 @@
+lightfm
+sklearn
+requests
```
[the remark this answers is not in the document]
Comment (TariqAHassan) [reply requirements.txt]: Good catch. This is used in `tests/`, but not in the broader package.

File: requirements.txt [hunk 4]
```diff
@@ -1,3 +1,6 @@
+lightfm
+sklearn
```
[the remark this answers is not in the document]
Comment (TariqAHassan) [reply requirements.txt]: With requiements.txt, I think there are three main options (but let me know if you have others):

1. replacing requirement.txt with `setup.py` (`i.e., the `install_requires` parameter). This way, users would sure to get dependencies required to run TensorRec.
2. keep requirements.txt and read it into `setup.py`. However it is my understanding that this is generally considered a bad practice (e.g., [here](http://alexanderwaldin.github.io/packaging-python-project.html)).
3. maintain both requirements.txt and `setup.py`

I think option one is best, but I'd be interested to you what you think.
If you do decide to go with option one, `requirements_dev.txt` could list everything requiements.txt currently does, but with some additional packages needed for development (like `requests`).

Lastly, `requirements_examples.txt` sounds like a good idea.

**Comment by TariqAHassan** [comment]
I've added a `requirements_examples.txt` file and restructured the other requirements files as described above.

Could `requirements_test.txt` simply become `requirements_dev.txt`?

**Comment by TariqAHassan** [comment]
After thinking about this PR, I think there has been a bit of 'mission creep' on my part. It was meant to only add documentation, but I also changed other things (such as setup.py). I am also unsure about the documentation style I used (could be a bit tedious to maintain).

So, I think I'll close this for now and aim to make distinct, more well-defined PRs later in the week.

**Pull request closed without merging**